A4A > Feedback: Implement updated Contextual Feedback Mechanism

yashwin opened the pull request:
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2006
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2007
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2008
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2009
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2010

## Proposed Changes

This PR:

- Updates the contextual feedback mechanism as the new requirements and design, including the adding referrals, inviting team member, updating agency details flows.
- Adds the feedback to the post-purchase flow

Note: Reset the preference (instructions below) before testing.

## Why are these changes being made?

* To help bolster our retention and get relevant, in-context feedback from agencies.

## Testing Instructions

* ~Patch 179607-ghe-Automattic/wpcom to your sandbox~
* Open the A4A live link
* Perform the below actions and verify the feedback is shown as displayed below. Fill out the form and submit it > Verify the form is submitted by checking the Slack message (channel name can be found here: https://github.com/Automattic/automattic-for-agencies-dev/issues/2006). 

1) Sending a referral (/marketplace/products > Toggle the refer mode on and send a referral)

<img width="1660" alt="Screenshot 2025-04-08 at 12 18 32 PM" src="https://github.com/user-attachments/assets/34acaca0-8d95-48c0-b7f0-57d199ed10b8" />

2) Adding a team member (/team/invite > Send invite)

<img width="1660" alt="Screenshot 2025-04-08 at 12 17 23 PM" src="https://github.com/user-attachments/assets/fab6d9bd-ba16-4600-84c4-e122d6f99cd2" />

3) Filling out partner directory information (/partner-directory/agency-details > Save public profile)

<img width="1666" alt="Screenshot 2025-04-04 at 10 54 36 AM" src="https://github.com/user-attachments/assets/b254b60f-2027-4ab4-9c57-de965dc16439" />

* Go to Marketplace > For the flows mentioned below: Verify the feedback form is shown > Submit the form > Verify you are redirected to the appropriate pages > Verify the form is submitted by checking the Slack message.

Note: Reset the preference (instructions below) after each flow.

1)  Go to Sites Dashboard > Click on Add + on backup or scan on any site > Purchase the product > Verify the feedback form is shown > Submit the form or skip it > Verify you are redirected to the sites dashboard
2) Go to Marketplace > Add a product (not free, non-WooCommerce) > Go to Checkout > Purchase the product > Verify you are redirected to the Licenses page > Click the `Assign` button > Assing the license to a site > Verify the feedback form is shown > Submit the form or skip it > Verify you are redirected to the Licenses page
3) Go to Marketplace > Add a product (not free, WooCommerce) > Go to Checkout > Purchase the product > Verify you are redirected to the Licenses page > Click the `Assign` button > Assing the license to a site > Verify that you are redirected to the Download Products page > Click the `Done` button > Verify the feedback form is shown > Submit the form or skip it > Verify you are redirected to the Licenses page
4) Go to Marketplace > Add a product (free, WooCommerce) > Go to Checkout > Purchase the product > Verify you are redirected to the Licenses page > Click the `Assign` button > Assing the license to a site > Verify that you are redirected to the Download Products page > Click the `Done` button > Verify the feedback form is not shown > Verify you are redirected to the Licenses page
3) Go to Markeplace > Add WP.com hosting > Go to Checkout > Purchase the hosting > Verify you are redirected to the Needs setup page > Create a site > Once you are in the Sites Dashboard > Click the `Set up your site` button on the banner > Verify you are redirected to the WP.com overview page > Go to A4A dashboard > Verify the feedback form is shown > Submit the form or skip it > Verify you are redirected to the Licenses page
4) Go to Markeplace > Add Pressable hosting > Go to Checkout > Purchase the hosting > Verify you are redirected to the Licenses page > Click the `Manage in Pressable` button > Verify you are redirected to Pressable > Go to A4A dashboard > Verify the feedback form is shown > Submit the form or skip it > Verify you are redirected to the Licenses page

<img width="1660" alt="Screenshot 2025-04-08 at 12 16 41 PM" src="https://github.com/user-attachments/assets/4193df0f-e847-4574-ad0b-547cd0e4ed1a" />

---
How the preferences are stored:

<img width="304" alt="Screenshot 2024-12-05 at 11 06 08 AM" src="https://github.com/user-attachments/assets/1ca12427-cee1-4f6c-9a3c-c40272e94c6a">

Preferences can be reset from staging
---
## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
update contextual feedback to support suggestions

yashwin committed:
implement contextal feedback for post-purchase flow

yashwin commented on client/a8c-for-agencies/lib/permission.ts:
```diff
@@ -113,7 +113,7 @@ const MEMBER_ACCESSIBLE_DYNAMIC_PATHS: Record< string, string[] > = {
 const DYNAMIC_PATH_PATTERNS: Record< string, RegExp > = {
 	'sites-overview': /^\/sites\/overview\/[^/]+(\/.*)?$/,
-	marketplace: /^\/marketplace\/[^/]+\/[^/]+(\/.*)?$/,
+	marketplace: /^\/marketplace\/[^/]+(\/[^/]+)?(\/.*)?(\?.*)?(#.*)?$/,
```
To support URLs required for feedback

yashwin commented on client/a8c-for-agencies/sections/marketplace/hosting-overview-v3/index.tsx:
```diff
@@ -136,7 +136,7 @@ function HostingOverviewV3( { section }: SectionProps ) {
 							onRemoveItem={ onRemoveCartItem }
 							onCheckout={ () => {
-								page( A4A_MARKETPLACE_CHECKOUT_LINK );
+								page.redirect( A4A_MARKETPLACE_CHECKOUT_LINK );
```
To make sure the pathname is updated accurately.

yashwin committed:
use values instead of text

yashwin committed:
minor ui fix

madebynoam reviewed (changes requested):
Thanks for this, @yashwin! 
Overall it looks good but we're missing a few visual details, uDy4bMwOiyDaQLiDeemNTv-fi-8657_5091.

1. The page BG should be #F8F8F8.
2. The icon buttons don't look like the latest pass. They should not have a fill. Just the outline if selected. And the colors should now be System colors.
3. The padding in of the form container looks off at the top.

madebynoam reviewed (approved):
@yashwin and I discussed that we'll tackle the updates to the emotion control in the next cycle, and use what we have currently to ship. And we'll tackle the page BG and padding in this release.

jkguidaven commented on client/a8c-for-agencies/components/a4a-feedback/index.tsx:
```diff
@@ -16,28 +16,34 @@ export interface Props extends FeedbackProps {
+	const [ suggestions, setSuggestions ] = useState< FeedbackSuggestion[] >( [] );
+
+	const onSuggestionChange = ( option: FeedbackSuggestion ) => {
+		if ( suggestions.find( ( suggestion ) => suggestion.value === option.value ) ) {
+			setSuggestions( ( prev ) =>
+				prev.filter( ( suggestion ) => suggestion.value !== option.value )
+			);
+		} else {
+			setSuggestions( ( prev ) => [ ...prev, option ] );
+		}
```
Nitpicking: So it is clearer what this line of code is doing.  

```
// Toggle suggestion selection: remove if already selected, add if not
		setSuggestions( ( prev ) => {
			const isSelected = prev.some( ( suggestion ) => suggestion.value === option.value );
			return isSelected
				? prev.filter( ( suggestion ) => suggestion.value !== option.value )
				: [ ...prev, option ];
		} );
```

jkguidaven commented on client/a8c-for-agencies/sections/marketplace/products-overview-v2/hooks/use-issue-and-assign-licenses.tsx:
```diff
@@ -235,7 +262,11 @@ function useIssueAndAssignLicenses(
 			}
 
 			// Otherwise, send them to the overview of all licenses
```
Can we move this closer to the `page.redirect` and not above the if condition?

yashwin committed:
fix the padding

yashwin commented:
Thanks for the review, @madebynoam!

I have fixed the padding issue. Regarding the background, as discussed, it is a bit tricky to achieve this. So, we will skip if for the initial push.

jkguidaven reviewed (approved):
Overall looks good to me. I left minor comments which are non-blocking.

jeffgolenski commented:
Thanks @yashwin! Aside from @madebynoam's feedback, I just have a couple of thoughts:

* All flows: After an action is conducted and we display the feedback prompt, can we hide the page header when the feedback prompt is visible? I'd like to keep the feedback prompt as focused as possible IE: For Referrals, can we hide "Your referrals and commissions"?
* For referrals specifically, can we also hide the notice to add payment methods until after the feedback prompt? We'll display that right after on the referrals dashboard, anyway. 
* For referrals: The sentence "Assign products to a site after they pay" feels a little strange in this location. Can we alter that to say "We emailed your referral order to [email]. You can assign your products to a site or set up hosting once they pay."

![CleanShot 2025-04-07 at 16 39 55](https://github.com/user-attachments/assets/a3964442-29b3-4b69-ab11-a8662f646423)

* Follow up for referrals UI: I also noticed that after I fill out and send the feedback, the referrals UI has a regression on wide screens. Some alignment issues—especially with the 4 cards above the table.

![CleanShot 2025-04-07 at 16 41 11](https://github.com/user-attachments/assets/c3cbef18-da26-40b3-a03e-e932197ce632)

### Purchase flow adjustments

* The mechanism works well, but doesn't display in the flow where it should. Details: pgjTho-dq-p2#comment-244. 

* We should only show this for non-free products and hosting. 

**Flow for paid products**

* It should display after license assignment for products. Example: User buys a Woo ext, assigns it, we show this download screen. Once they click "done" we take them to /purchases/licenses and that's when we should show it. 

![CleanShot 2025-04-07 at 17 00 21](https://github.com/user-attachments/assets/b75017c6-124e-47db-b005-e189a4178b10)

**Flow for hosting**

Here's a video to explain what I mean! Let me know if it's too challenging. 

https://github.com/user-attachments/assets/b03d5275-1d08-4f93-8ec6-1d080435ec49



* The prompts should only display after a hosting product is "set up".  We want to ensure we're capturing feedback to deliver to WordPress.com and Pressable, as well. 

* For WordPress.com, after they click "Create site" and configure things, we begin to spin up an instance. Then we ask them to set up their site. After they set up their site in a new tab, they should return to A4A. That's when we should display the feedback prompt. 
* Same for Pressable plan set ups, respectfully.

yashwin committed:
create feedback section and update logic

yashwin commented:
Thanks for the review & feedback, everyone!

There is a significant change in the logic and feedback is rendered now. So, could you please do a fresh review? I have updated the testing instructions according to the new changes:

Updates are done as per @jeffgolenski's comment:

> All flows: After an action is conducted and we display the feedback prompt, can we hide the page header when the feedback prompt is visible? I'd like to keep the feedback prompt as focused as possible IE: For Referrals, can we hide "Your referrals and commissions"?

> For referrals specifically, can we also hide the notice to add payment methods until after the feedback prompt? We'll display that right after on the referrals dashboard, anyway.

Updated the way the feedback is rendered as per @jkguidaven's suggestion and we now have a separate page for the feedback. This makes it easy to remove the page header and the banners.

> For referrals: The sentence "Assign products to a site after they pay" feels a little strange in this location. Can we alter that to say "We emailed your referral order to [email]. You can assign your products to a site or set up hosting once they pay."

Done. 

> Follow up for referrals UI: I also noticed that after I fill out and send the feedback, the referrals UI has a regression on wide screens. Some alignment issues—especially with the 4 cards above the table.

Fixed 👍 

> The mechanism works well, but doesn't display in the flow where it should. Details: pgjTho-dq-p2#comment-244.
> We should only show this for non-free products and hosting.
> Flow for paid products
> It should display after license assignment for products. Example: User buys a Woo ext, assigns it, we show this download screen. Once they click "done" we take them to /purchases/licenses and that's when we should show it.

Done

> Flow for hosting
> The prompts should only display after a hosting product is "set up". We want to ensure we're capturing feedback to deliver to WordPress.com and Pressable, as well.
> For WordPress.com, after they click "Create site" and configure things, we begin to spin up an instance. Then we ask them to set up their site. After they set up their site in a new tab, they should return to A4A. That's when we should display the feedback prompt.
> Same for Pressable plan set ups, respectfully.

All done.

jkguidaven reviewed (approved):
Thanks for applying my suggestion. Retested some flow and didn't find any issues. The changes also LGTM. Nice work! 👏

yashwin committed:
update survey keys to support MC tool

jeffgolenski commented:
Just tested again @yashwin. Awesome work. The stand-alone page for the feedback prompt is great! The prompt also now displays properly after hosting set up. SO GOOD. Thank you. Two last things I noticed:

* The "Thank you, we'll use this feedback to prioritize..." notice seems to disappear really quickly. Can we adjust that timeframe a bit so users have a moment to read it?
* For maintenance outside of this project: The referral header seems to have a regression. I seem to a recall a primary CTA in that header. "Send a new referral" that would take the user to the marketplace with referral mode toggled on. Did I imagine that? 😆  It's gone now. cc @madebynoam 

![CleanShot 2025-04-09 at 12 06 54](https://github.com/user-attachments/assets/86cf4e84-6ea3-4c66-b51f-0fca448fc701)

jeffgolenski commented on client/a8c-for-agencies/components/a4a-feedback/style.scss:
```diff
@@ -8,17 +8,24 @@ $highlight-color: #3858E9;
 
 .a4a-feedback__content {
-	padding-block-start: 32px;
+	margin-block-start: calc(48px + 8px);
```
Look at you, gettin' fancy with calc!

yashwin committed:
update notice duration

yashwin commented:
> Just tested again @yashwin. Awesome work. The stand-alone page for the feedback prompt is great! The prompt also now displays properly after hosting set up. SO GOOD. Thank you. Two last things I noticed:

Thank you, Jeff!

> * The "Thank you, we'll use this feedback to prioritize..." notice seems to disappear really quickly. Can we adjust that timeframe a bit so users have a moment to read it?

Yes, I made it 10 seconds before shipping!

> * For maintenance outside of this project: The referral header seems to have a regression. I seem to a recall a primary CTA in that header. "Send a new referral" that would take the user to the marketplace with referral mode toggled on. Did I imagine that? 😆  It's gone now. cc @madebynoam

This is hidden mostly because the payout method has not been added. @jkguidaven, IIRC, we enabled Referrals even if the payout method is not added.
